core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -126,6 +127,8 @@
         defaultRequestObjectResolvers = builder.build();
     }
 
+    private static final Splitter commaDelimitedQueryParamSplitter = Splitter.on(',').trimResults();
```

Comment:
Should we modify `QueryParams.fromQueryString(queryString)` rather than fix it only in annotated service?

Reply:
> Can we modify the decoder?

Sure. 🙂 If you need to modify it, you can do it.
One remaining concern is a breaking change. If we change the default behavior, I guess, it could be an unexpected change for some users.
Do you think this could be the default behavior?
Otherwise, should we add a flag like `Flags.allowCommaDelimitedQueryParameters()` and let users turn on the feature manually? If so, I think we can turn on it by default in Armeria 2.0.

Reply:
For the sake of convenience of Spring users, we may add the property to `ArmeriaSettings`

Reply:
Because it's not a standard, + 1 for adding Flags.

Reply:
Because the `QueryParams` doesn't know the target class to resolve, it will always delimit query parameters if contain the comma. That is, we should delimit `@Param List<String> foo` but shouldn't `@Param String foo`. But Armeria will resolve the first element if the corresponding value contains the comma.
So, modifying `QueryParams` isn't suitable I guess.

Reply:
Good point. It could be vague and cause extra confusion. 
It might be clear to only apply support for comma-delimited to annotated services.
Optionally, if users don't want to turn on the flag (`Flags.allowCommaDelimitedQueryParameters()`) globally, we might provide `@Delimited(",")` you said in https://github.com/line/armeria/pull/4060#discussion_r806616127.

Reply:
Yeah, the flag looks good to me also 👍 We can introduce a kind of `@Delimited` when we find a better solution.

BTW, how can I test using the flag...? Would it be better to add `allowCommaDelimitedQueryParameters` method that has `Flags.allowCommaDelimitedQueryParameters()` by default to `AnnotatedServiceBindingBuilder`?

Reply:
> Would it be better to add allowCommaDelimitedQueryParameters method that has Flags.allowCommaDelimitedQueryParameters() by default to AnnotatedServiceBindingBuilder?

That's a good idea. 👍

Reply:
Our flags were so far mostly for changing some thresholds (timeouts, cache sizes, ..). I agree that a user should be given some options for convenience, but are we sure we want to introduce a flag for that? How about just starting from introducing an additional parameter to `QueryParams.fromQueryString()` and `QueryParams.toQueryString()`? e.g.

```java
p = QueryParams.fromQueryString("a=1;a=2,3", QueryStringDecoderOptions.SEMICOLON_AS_SEPARATOR, QueryStringDecoderOptions.COMMA_SOMETHING_SOMETHING)

assert(
  "a=1,2,3".equals(
    p.toQueryString(QueryStringEncoderOptions.COMMA_SOMETHING_SOMETHING)
  )
)
```

Reply:
I introduced `useQueryDelimiter` for building an annotated service:

```java
sb.annotatedService()
  .useQueryDelimiter(",")
  .build(new MyAnnotatedService());
```

Actually, we don't have to introduce a new flag because it makes sense that not specifying `useQueryDelimiter` means not allowing delimition.  WDYT?

Reply:
> Actually, we don't have to introduce a new flag because it makes sense that not specifying useQueryDelimiter means not allowing delimition. WDYT?

I can't speak for @trustin on his original intention, but I agree with your statement.
This feature is `annotationService` specific, so I guess it defining as a flag is unnecessary.
I guess just remove the `Flags.java` implementation and by default `queryDelimiter = null` makes sense 👍